go-controller/pkg/clustermanager/userdefinednetwork/controller.go

```diff
@@ -335,12 +342,192 @@ func (c *Controller) reconcileCUDN(key string) error {
 	return errors.Join(syncErr, updateStatusErr)
 }
 
-func (c *Controller) syncClusterUDN(_ *cudnv1.ClusterUserDefinedNetwork) ([]netv1.NetworkAttachmentDefinition, error) {
-	// TODO: implement
-	return nil, fmt.Errorf("implement me")
+func (c *Controller) syncClusterUDN(cudn *cudnv1.ClusterUserDefinedNetwork) ([]netv1.NetworkAttachmentDefinition, error) {
+	if cudn == nil {
+		return nil, nil
+	}
+
+	selectedNamespaces, err := c.getSelectedNamespaces(cudn.Spec.NamespaceSelector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get selected namespaces: %w", err)
+	}
+	affectedNamespaces := c.namespaceTracker[cudn.Name]
+
+	if !cudn.DeletionTimestamp.IsZero() {
+		if controllerutil.ContainsFinalizer(cudn, template.FinalizerUserDefinedNetwork) {
+			var errs []error
+			for nsToDelete := range affectedNamespaces.Union(selectedNamespaces) {
```

Comment:
I don't think we need to account for selectedNamespaces here, since all namespace that had NADs created should be in the affectedNamespaces, and even if new namespaces are selected now, we have nothing to clean up for them?

Reply:
Initially I tried to go on the safe side and include the selected-namespaces, just in case.
Looking at this again it make sense to go over the affected namespaces only.
Thanks for the heads up :slightly_smiling_face: